### PR TITLE
update go-apidiff

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

This PR fixes some false-positives that make the apidiff CI job report failure for some changes. Bumping the version to v0.8.2 seems to fix some generics-related issues. The script change only checks the output for `api` package names when they appear at the beginning of a line to prevent things like API types being used as function parameters from being flagged.

⚠️ WARNING: If you'd like to test this locally, do so either in a fresh clone or in a separate [worktree](https://git-scm.com/docs/git-worktree) because `go-apidiff` will run `git clean` under the hood and delete all your gitignored files like `tilt-settings.yaml` (https://github.com/joelanford/go-apidiff/issues/62).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by mboersma — APPROVED** [review]
/lgtm

**Review by willie-yao — COMMENTED** [review]
/lgtm

File: scripts/ci-apidiff.sh [hunk 1]
```diff
@@ -24,7 +24,7 @@ cd "${REPO_ROOT}" || exit
 echo "*** Running go-apidiff ***"
 APIDIFF=$(APIDIFF_OLD_COMMIT="${PULL_BASE_SHA}" make apidiff 2> /dev/null)
 
-if [[ "${APIDIFF}" == *"sigs.k8s.io/cluster-api-provider-azure/api/"* ]] || [[ "${APIDIFF}" == *"sigs.k8s.io/cluster-api-provider-azure/exp/api/"* ]]; then
+if grep -qE '^sigs\.k8s\.io/cluster-api-provider-azure/(exp/)?api/' <<< "$APIDIFF"; then
```
Comment (jackfrancis) [inline scripts/ci-apidiff.sh:27]: I know this isn't a functional change but can we dig up documentation (inherited from k8s standards is sufficient) that outlines that changing *any* API classified as experimental is permitted? I am uncertain if, say, a v1beta1 API is breakable even if it's experimental + feature flag.

**Review by jackfrancis — APPROVED** [review]
/lgtm
/approve

File: scripts/ci-apidiff.sh (hunk 1 above)
Comment (nojnhuh) [reply scripts/ci-apidiff.sh:27]: I think this is the relevant guidance:
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md

I don't see any mention of what's expected when an API and an accompanying feature have differing alpha/beta/GA status, so I think the API-specific guidance is most relevant here. It looks to me like both alpha and beta APIs can make breaking changes between versions, but for beta APIs we need to automate or document how to upgrade.

Comment (jackfrancis) [reply scripts/ci-apidiff.sh:27]: Thanks, I think we're good here!

**Pull request merged**